### ISSUE TITLE
Syntax errors lead to unclosed R2DBC connection in SQL Server

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14489]
The following test is failing:

```java
assertEquals(-1,
    Flux.from(selectOne())
        .map(r -> 1)
        .onErrorReturn(DetachedException.class, -1)
        .blockFirst());
```

Note that the `selectOne()` query is created statically, so a `DetachedException` is thrown:

```
08:48:13,949 DEBUG [LoggerListener                ] - Exception                
org.jooq.exception.DetachedException: Cannot execute query. No JDBC Connection configured
	at org.jooq_3.18.0-SNAPSHOT.DEFAULT.debug(Unknown Source) ~[?:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:310) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:301) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:322) ~[classes/:?]
	at org.jooq.impl.R2DBC$BlockingRecordSubscription.request0(R2DBC.java:1470) ~[classes/:?]
	at org.jooq.impl.ThreadGuard.lambda$0(ThreadGuard.java:74) ~[classes/:?]
	at org.jooq.impl.ThreadGuard.run(ThreadGuard.java:87) ~[classes/:?]
	at org.jooq.impl.ThreadGuard.run(ThreadGuard.java:74) ~[classes/:?]
	at org.jooq.impl.R2DBC$AbstractSubscription.request(R2DBC.java:185) ~[classes/:?]
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:164) ~[reactor-core-3.4.19.jar:3.4.19]
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:2194) ~[reactor-core-3.4.19.jar:3.4.19]
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onSubscribe(FluxOnErrorResume.java:74) ~[reactor-core-3.4.19.jar:3.4.19]
	at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:92) ~[reactor-core-3.4.19.jar:3.4.19]
	at org.jooq.impl.ResultQueryTrait.subscribe(ResultQueryTrait.java:372) ~[classes/:?]
	at org.jooq.impl.SelectImpl.subscribe(SelectImpl.java:2851) ~[classes/:?]
	at reactor.core.publisher.FluxSource.subscribe(FluxSource.java:67) ~[reactor-core-3.4.19.jar:3.4.19]
	at reactor.core.publisher.Flux.subscribe(Flux.java:8466) ~[reactor-core-3.4.19.jar:3.4.19]
	at reactor.core.publisher.Flux.blockFirst(Flux.java:2599) ~[reactor-core-3.4.19.jar:3.4.19]
	at org.jooq.test.all.testcases.R2DBCTests.testR2DBCException(R2DBCTests.java:701) ~[test-classes/:?]
	at org.jooq.test.jOOQAbstractTest.testR2DBCException(jOOQAbstractTest.java:2890) ~[test-classes/:?]
```

However, in the SQL Server test suite, our connection lifecycle check now fails:

```
java.lang.AssertionError: expected:<0> but was:<2>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at org.jooq.test.util.BaseTest.assertEquals(BaseTest.java:275)
	at org.jooq.test.jOOQAbstractTest.tearDown(jOOQAbstractTest.java:837)
```

The test is:

```java
// [#13343] [#13669] [#13738] R2DBC connections must always be closed
assertEquals(0, dbSetup.openR2dbcConnectionCounter.get());
```

I feel this is a regression from recent changes, possibly not released yet (?). To be verified. Only SQL Server seems to be affected.

**Comment by lukaseder** [comment]
The test fails also on the 3.17.0 branch, so the recent upgrade to R2DBC `1.0.0.RELEASE` isn't the problem here: https://github.com/jOOQ/jOOQ/issues/13565

**Comment by lukaseder** [comment]
It's really just the test with the `DetachedException` that causes issues. The bad grammar test doesn't:

```java
assertEquals(-1,
    Mono.from(r2dbc().query("invalid sql"))
        .doOnError(DataAccessException.class, e -> {
            assertEquals(DataAccessException.class, e.getClass());
            assertTrue(R2dbcException.class.isInstance(e.getCause(R2dbcException.class)));
            // [#12199] TODO: Support translating these
            // assertEquals(C42_SYNTAX_ERROR_OR_ACCESS_RULE_VIOLATION, e.sqlStateClass());
            assertTrue(e.getMessage(), e.getMessage().contains("invalid sql"));
        })
        .onErrorReturn(DataAccessException.class, -1)
        .block());
```

A hint is in the stack trace: We're running the blocking execution path:

```
08:48:13,949 DEBUG [LoggerListener                ] - Exception                
org.jooq.exception.DetachedException: Cannot execute query. No JDBC Connection configured
	at org.jooq_3.18.0-SNAPSHOT.DEFAULT.debug(Unknown Source) ~[?:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:310) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:301) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:322) ~[classes/:?]
	at org.jooq.impl.R2DBC$BlockingRecordSubscription.request0(R2DBC.java:1470) ~[classes/:?]
```

Which seems reasonable at first, because the query is detached. And the `R2DBC` path doesn't ever throw a `DetachedException`, always assuming a `io.r2dbc.Connnection` is present for the `R2DBC` path to apply

**Comment by lukaseder** [comment]
There might be a temptation to fail early, already in this method:

```
	at org.jooq.impl.ResultQueryTrait.subscribe(ResultQueryTrait.java:372) ~[classes/:?]
```

But this would prevent supplying a `Connection` from within an `ExecuteListener`, so not an option in the long run. (Though currently, `ExecuteListener` aren't supported with R2DBC execution yet: https://github.com/jOOQ/jOOQ/issues/12221)

**Comment by lukaseder** [comment]
OK, I was mislead. The path that goes through the blocking execution is fine. The problems are really the queries where the SQL is incorrect. They leave open connections.

**Comment by lukaseder** [comment]
The problem is that we're never completing the execution in `R2DBC`:

```java
@Override
public final void onError(Throwable t) {
    resultSubscriber.downstream.subscriber.onError(translate(resultSubscriber.downstream.sql(), t));
    complete(true);
}
```

A `try-catch` around the delegating call doesn't help (?)

**Comment by lukaseder** [comment]
Welp. When setting breakpoints, then the tests also fail in PostgreSQL. Let this be again a stark reminder how brittle everything in this reactive world is 🤨

**Comment by lukaseder** [comment]
Perhaps this is just another instance of https://github.com/jOOQ/jOOQ/issues/12977 ? In that issue, only the `onComplete()` signals have been re-sequenced, not the `onError()` signals!

**Comment by lukaseder** [comment]
Yes, #12977 was wrong as well, but I think I've finally found the main issue to all of these problems! The `Forwarding` subscriber skips one level of `Publisher` nesting:

- `ResultQuery` acquires a
  - `Connection` executes a
    - `Statement` produces a set of (!)
      - `Result` produces a set of
        - `Long` (via `Result::getRowsUpdated`) or `Record` (via `Result::map`)

Observe the fact that a `Statement` may produce several `Result`. There are at least 2 cases where this is relevant:

- Batch statements (multiple update counts come in separate `Result`, at least in some drivers, not in separate `Long`
- Statement batches, i.e. strings containing several statements may produce the `Record` result set we're interested in only in the second `Result`, the first one being an update count.

One example of the latter is in SQL Server:

```sql
declare @result table ([id] int);
-- This produces an update count, which we must skip
insert into [test].[dbo].[t_identity_pk] ([val]) output inserted.[id] into @result values (1);
-- This produces a set of rows
select [id] from @result [r];
```

Somehow, the `Forwarding` subscriber tried to implement both 2 lowest levels of nested `Publisher`, when in fact we must separate the two levels.

**Comment by lukaseder** [comment]
RS is such a mess 😅 (I'm documenting all this here, in case I need it again in the future). So, the `Forwarding` subscriber did things in such complicated ways to correctly implement the RS TCK. A simpler alternative fixes the integration tests, but fails the TCK (e.g. `required_spec105_mustSignalOnCompleteWhenFiniteStreamTerminates`)

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0 and 3.17.7 (https://github.com/jOOQ/jOOQ/issues/14493)
